src/OVAL/probes/unix/linux/auditdline_probe.c

```diff
@@ -0,0 +1,1077 @@
+	int i, retval;
+	int line_number = 1;
+	int timeout = 40; /* loop has delay of .1 - so this is 4 seconds */
+	struct audit_reply rep;
+	fd_set read_mask;
+	FD_ZERO(&read_mask);
+	FD_SET(fd, &read_mask);
+
+	for (i = 0; i < timeout; i++) {
+		struct timeval t;
+
+		t.tv_sec = 0;
+		t.tv_usec = 100000; /* .1 second */
+		do {
+			retval = select(fd + 1, &read_mask, NULL, NULL, &t);
+		} while (retval < 0 && errno == EINTR);
+		// We'll try to read just in case
+		retval = audit_get_reply(fd, &rep, GET_REPLY_NONBLOCKING, 0);
+		if (retval > 0) {
+			if (rep.type == NLMSG_ERROR && rep.error->error == 0) {
+				i = 0; /* reset timeout */
+				continue; /* This was an ack */
+			} else if (rep.type == NLMSG_DONE) {
+				break;
+			} else if (rep.type == NLMSG_ERROR) {
+				int errnum = rep.error->error;
+				dE("Error while reading audit rules packets from the kernel : %d",
+				   errnum);
+				SEXP_t *item =
+					probe_item_create(OVAL_LINUX_AUDITDLINE,
+							  NULL, "filter_key",
+							  OVAL_DATATYPE_STRING,
+							  "", NULL);
+
+				probe_item_setstatus(item,
+						     SYSCHAR_STATUS_ERROR);
+				probe_item_add_msg(
+					item, OVAL_MESSAGE_LEVEL_ERROR,
+					"Error while reading audit rules packets from the kernel : %d",
+					errnum);
+				probe_item_collect(ctx, item);
+			}
+			struct audit_line *audit_line = get_rule(rep.ruledata);
+			audit_line->line_number = line_number;
+			oscap_list_add(audit_list, audit_line);
+			line_number++;
+		}
+	}
+	return audit_list;
+}
+
+// End of the modified code
+
+static struct oscap_list *get_all_audit_rules(probe_ctx *ctx)
+{
+	int audit_fd = audit_open();
+
+	if (audit_fd > -1) {
+		// Request the kernel to provide the audit rules list
+		if (audit_request_rules_list_data(audit_fd) > 0) {
+			dD("Successfull request to get the kernel audit rules.");
+
+			struct oscap_list *audit_list =
+				get_reply(audit_fd, ctx);
+			audit_close(audit_fd);
+			return audit_list;
+		} else {
+			int errnum = errno;
+			dE("Failed to request the kernel for audit rules : %s",
+			   strerror(errnum));
+			SEXP_t *item = probe_item_create(OVAL_LINUX_AUDITDLINE,
+							 NULL, "filter_key",
+							 OVAL_DATATYPE_STRING,
+							 "", NULL);
+
+			probe_item_setstatus(item, SYSCHAR_STATUS_ERROR);
+			probe_item_add_msg(
+				item, OVAL_MESSAGE_LEVEL_ERROR,
+				"Failed to request the kernel for audit rules : %s.",
+				strerror(errnum));
+			probe_item_collect(ctx, item);
+		}
+	} else {
+		int errnum = errno;
+		dE("Failed to get a handle for the audit system : %s",
+		   strerror(errnum));
+		SEXP_t *item =
+			probe_item_create(OVAL_LINUX_AUDITDLINE, NULL,
+					  "filter_key", OVAL_DATATYPE_STRING,
+					  "", NULL);
+
+		probe_item_setstatus(item, SYSCHAR_STATUS_ERROR);
+		probe_item_add_msg(
+			item, OVAL_MESSAGE_LEVEL_ERROR,
+			"Failed to obtain a handle to the audit system : %s.",
+			strerror(errnum));
+		probe_item_collect(ctx, item);
+	}
+
+	audit_close(audit_fd);
```

Comment:
This audit_close should be moved just above the else because that is the only time the audit_fd descriptor is valid.